contracts/storage/ServiceAgreementStorage.sol

```diff
@@ -0,0 +1,410 @@
+                    msg.sender,
+                    maxProofWindowOfssetPerc - minProofWindowOfssetPerc + 1
+                );
+            }
+
+            serviceAgreements[agreementId].epochSubmissionHeads[epoch] = commitId;
+            commitSubmissions[commitId] = commit;
+            _link_commits(agreementId, epoch, commit.identityId, prevHeadIdentityId);
+        }
+        else {
+            bytes32 prevCommitId = keccak256(abi.encodePacked(agreementId, epoch, prevIdentityId));
+            CommitSubmission memory prevCommit = commitSubmissions[prevCommitId];
+
+            require(
+                commit.score <= prevCommit.score,
+                "Score of the commit must be less or equal to the one you want insert after!"
+            );
+
+            uint96 nextIdentityId = prevCommit.nextIdentity;
+            if (nextIdentityId != 0) {
+                bytes32 nextCommitId = keccak256(abi.encodePacked(agreementId, epoch, nextIdentityId));
+                CommitSubmission memory nextCommit = commitSubmissions[nextCommitId];
+
+                require(
+                    commit.score >= nextCommit.score,
+                    "Score of the commit must be greater or equal to the one you want insert before!"
+                );
+            }
+
+            commitSubmissions[commitId] = commit;
+            _link_commits(agreementId, epoch, prevIdentityId, commit.identityId);
+            _link_commits(agreementId, epoch, commit.identityId, nextIdentityId);
+        }
+    }
+
+    function _link_commits(bytes32 agreementId, uint16 epoch, uint96 leftIdentityId, uint96 rightIdentityId)
+        private
+    {
+        bytes32 leftCommitId = keccak256(abi.encodePacked(agreementId, epoch, leftIdentityId));
+        commitSubmissions[leftCommitId].nextIdentity = rightIdentityId;
+    }
+
+    function _generateAgreementId(address assetTypeContract, uint256 tokenId, bytes keyword, uint8 hashingAlgorithm)
+        private
+        returns (bytes32)
+    {
+        HashingHub hashingHub = HashingHub(hub.getContractAddress("HashingHub"));
+        return hashingHub.callHashFunction(hashingAlgorithm, abi.encodePacked(assetTypeContract, tokenId, keyword));
+    }
+
+    function _calculateScore(uint8 hashingAlgorithm, bytes nodeId, bytes keyword, uint32 stake, uint32 a, uint32 b)
+        private
+        returns (uint32)
+    {
+        ShardingTable shardingTable = ShardingTable(hub.getContractAddress("ShardingTable"));
+        // TODO: Add function to the ShardingTable
+        bytes32 nodeIdHash = shardingTable.getNodeIdHash(nodeId, hashingAlgorithm);
+
+        HashingHub hashingHub = HashingHub(hub.getContractAddress("HashingHub"));
+        bytes32 keywordHash = hashingHub.callHashFunction(hashingAlgorithm, keyword);
+
+        uint32 distance = uint32(nodeIdHash ^ keywordHash);
+
+        // TODO: Change formula
+        return (a * stake) / (b * distance);
```

Comment:
split into separate ScoringFunctionHub contract so different scoring functions can be applied

Reply:
I replaced whole function with ScoringFunctionHub, because it will give us more flexibility in terms of choosing how to calculate distance etc.